chips/nrf52/src/ficr.rs

```diff
@@ -127,6 +127,8 @@ register_bitfields! [u32,
     ],
     /// Part code
     InfoPart [
+        // clippy allow needed for same reason overflowing_literals needed below
+        #[allow(clippy::enum_clike_unportable_variant)]
```

Comment:
Is there actually a reason for the `Unspecified` variant? IIRC this doesn't exist in the data sheet, so we could remove it, and instead use `Option<Enum>` to indicate an unspecified value (i.e. acceptable values will be `None`, and `Some(N528xx)` for the relevant `xx`).

cc @jmichelp.

Reply:
The `Unspecified` values are only used when converting between `u32` and the `Enum`. I think it's just old-style code and this could be replaced by implementing the `TryFrom` and `TryInto` traits IMO.

Reply:
But they actually are mentioned in the product specification BTW :)

Reply:
Given that the datasheet does list the `Unspecified` variants, it seems a little weird to remove them from the `register_bitfields()` call. But I am happy to make that change if people agree that is the preferred approach.

Reply:
For low-level stuff like this, I generally prefer matching what the hardware actually does (not that consumers are ever likely to have hands-on of a test chip set to this value, but nonetheless I don't think it's a value we are at liberty to specify).

---

Partially related, from the clippy source:
```rust
    /// **What it does:** Checks for C-like enumerations that are
    /// `repr(isize/usize)` and have values that don't fit into an `i32`.
    ///
    /// **Why is this bad?** This will truncate the variant value on 32 bit
    /// architectures, but works fine on 64 bit.
```

I don't understand why a **usize** or **u32** would be truncated.

Reply:
Yeah, I am continuing to dig into this and it seems incorrect that we labeled these with `#[allow(overflowing_literals)]` in the first place. The issue is that Rust is parsing these values as isize`, not `u32`.

Removing the overflowing_literalls allow leads to the following warning:
```
error: literal out of range for isize
   --> chips/nrf52/src/ficr.rs:140:27
    |
140 |             Unspecified = 0xffffffff
    |                           ^^^^^^^^^^
    |
    = note: `#[deny(overflowing_literals)]` on by default
    = note: the literal `0xffffffff` (decimal `4294967295`) does not fit into an `isize` and will become `-1isize`

error: aborting due to previous error

error: could not compile `nrf52`.
```

Is this an error in the tock-register-interface crate?

Reply:
Okay, I have convinced myself this is not an error. I extracted the relevant portions of the tock-register-interface macros as the following rust code:

```rust
pub trait TryFromValue<V> {
    type EnumType;

    fn try_from(v: V) -> Option<Self::EnumType>;
}

#[repr(i32)] // This is not included in the macro, instead the enum defaults to the default isize representation, which on our devices is i32.
#[derive(Debug)]
pub enum Value {
    First = 0,
    #[allow(overflowing_literals)]
    Second = 0xffffffff,
}

impl TryFromValue<u32> for Value {
    type EnumType = Value;

    fn try_from(v: u32) -> Option<Self::EnumType> {
        match v {
            x if x == Value::First as u32 => Some(Value::First),
            x if x == Value::Second as u32 => Some(Value::Second),
            _ => Option::None,
        }
    }
}

fn main() {
    let tmp = Value::try_from(0xffffffff);
    println!("tmp: {:?}", tmp);
    // works! tmp == Some(Second)
}
```

Without the overflowing literals allow, this wont compile, but with it, the `try_from` call is interpreted correctly. 

The basic problem is that the enum defined here: https://github.com/tock/tock/blob/master/libraries/tock-register-interface/src/macros.rs#L90-L113 uses the default representation (`isize`). This specific problem can be fixed by annotating the enum with `repr($valtype)`, but then the macro fails when passed zero-variant enums. It is probably possible to fix this problem within the macro so that the alternative representation is only used when the enum has 1+ variants, but given that regardless of the enum representation the registers will be parsed correctly, it seems more correct to annotate the enum *in the register_bitmasks macro* with `allow[overflowing_literals]` and `#[allow(clippy_enum_clike_unportable_variant)]`, so that users of the interface don't have to worry about this non-issue.

Reply:
actually, that is also not perfect, because it allows passing a value greater than the maximum usize...